promgen/tests/test_rest.py

```diff
@@ -21,3 +22,51 @@ def test_alert(self):
         response = self.fireAlert()
         self.assertEqual(response.status_code, 202)
         self.assertCount(models.Alert, 1, "Alert Queued")
+
+    @override_settings(PROMGEN=tests.SETTINGS)
+    def test_retrieve_farm(self):
+        response = self.client.get(reverse("api:farm-list"))
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 1)
+        self.assertEqual(response.data[0]["id"], 1)
+        self.assertEqual(response.data[0]["name"], "test-farm")
+        self.assertEqual(response.data[0]["source"], "promgen")
+        self.assertEqual(response.data[0]["url"], "http://promgen.example.com/farm/1")
+
+        response = self.client.get(reverse("api:farm-list"), {"name": "test-farm"})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 1)
+
+        response = self.client.get(reverse("api:farm-list"), {"source": "promgen"})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 1)
+
+        response = self.client.get(reverse("api:farm-list"), {"name": "farm", "name_exact_match": "false"})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 1)
+
+        response = self.client.get(reverse("api:farm-list"), {"source": "other-source"})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 0)
+
+        response = self.client.get(reverse("api:farm-list"), {"name": "farm", "name_exact_match": "true"})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 0)
+
+        response = self.client.get(reverse("api:farm-list"), {"name": "farm"})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 0)
+
+
+        farm = models.Farm.objects.get(id=1)
+        models.Host.objects.create(name="host.example.com", farm=farm)
+
+        response = self.client.get(reverse("api:farm-detail", args=[1]))
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data["id"], 1)
+        self.assertEqual(response.data["name"], "test-farm")
+        self.assertEqual(response.data["source"], "promgen")
+        self.assertEqual(response.data["url"], "http://promgen.example.com/farm/1")
+        self.assertEqual(len(response.data["hosts"]), 1)
+        self.assertEqual(response.data["hosts"][0]["url"], "http://promgen.example.com/host/host.example.com")
+        self.assertEqual(response.data["hosts"][0]["name"], "host.example.com")
```

Comment:
```suggestion

    @override_settings(PROMGEN=tests.SETTINGS)
    def test_retrieve_farm(self):
        expected = tests.Data("examples", "rest.farm.json").json()

        response = self.client.get(reverse("api:farm-list"))
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.json(), expected)

        response = self.client.get(reverse("api:farm-list"), {"name": "test-farm"})
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.json(), expected)

        response = self.client.get(reverse("api:farm-list"), {"source": "promgen"})
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.json(), expected)

        response = self.client.get(reverse("api:farm-list"), {"name": "other-name"})
        self.assertEqual(response.status_code, 200)
        self.assertEqual(len(response.data), 0)

        response = self.client.get(reverse("api:farm-list"), {"source": "other-source"})
        self.assertEqual(response.status_code, 200)
        self.assertEqual(len(response.data), 0)


        farm = models.Farm.objects.get(id=1)
        models.Host.objects.create(name="host.example.com", farm=farm)
        expected = tests.Data("examples", "rest.farm.1.json").json()

        response = self.client.get(reverse("api:farm-detail", args=[1]))
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.json(), expected)
```

- I have removed the part where we test the `name_exact_match`.
- I have added a test using a farm name that does not exist.
- We can make this test less verbose if we create the following JSON files with the expected responses:

"promgen/promgen/tests/examples/rest.farm.json"

```
[
  {
    "id": 1,
    "url": "http://promgen.example.com/farm/1",
    "name": "test-farm","source":"promgen"
  }
]
```

"promgen/promgen/tests/examples/rest.farm.1.json"

```
{
  "id": 1,
  "url": "http://promgen.example.com/farm/1",
  "name": "test-farm",
  "source": "promgen",
  "hosts": [
    {
      "name": "host.example.com",
      "url": "http://promgen.example.com/host/host.example.com"
    }
  ]
}
```

Also, it would be nice if you add a comment above each GET request explaining what are you testing. For instance:

```
        # Check retrieving all farms whose "source" is "promgen"
        response = self.client.get(reverse("api:farm-list"), {"source": "promgen"})

        [...]

        # Check retrieving farms with a non-existent "name" returns an empty list
        response = self.client.get(reverse("api:farm-list"), {"name": "other-name"})
```